packages/hawtio/src/plugins/shared/attributes/AttributeTable.tsx

```diff
@@ -0,0 +1,88 @@
+import { useContext, useEffect, useState } from 'react'
+import { Card, CardBody, Text } from '@patternfly/react-core'
+import { Table, TableBody, TableHeader, TableProps } from '@patternfly/react-table'
+import { PluginNodeSelectionContext } from '@hawtiosrc/plugins/selectionNodeContext'
+import { AttributeValues } from '@hawtiosrc/plugins/connect/jolokia-service'
+import { attributeService } from './attribute-service'
+import './AttributeTable.css'
+import { JmxContentMBeans } from '@hawtiosrc/plugins/shared/JmxContentMBeans'
+import { humanizeLabels } from '@hawtiosrc/util/strings'
+
+export const AttributeTable: React.FunctionComponent = () => {
+  const { selectedNode } = useContext(PluginNodeSelectionContext)
+  const [attributes, setAttributes] = useState<AttributeValues[]>([{}])
+  const [isReading, setIsReading] = useState<boolean>(false)
+
+  useEffect(() => {
+    if (!selectedNode) {
+      return
+    }
+
+    const childrenMbeansAttributes: AttributeValues[] = []
+    const readAttributes = async () => {
+      setIsReading(true)
+      if (selectedNode.children)
+        for (const mbean of selectedNode.children) {
+          const objectName = mbean.objectName
+          if (objectName) {
+            const attrs = await attributeService.read(objectName)
+            childrenMbeansAttributes.push(attrs)
+          }
+        }
+      setAttributes([...childrenMbeansAttributes])
+      setIsReading(false)
+    }
+    readAttributes()
+  }, [selectedNode])
+
+  if (!selectedNode) {
+    return null
+  }
+
+  if (isReading) {
+    return (
+      <Card>
+        <CardBody>
+          <Text component='p'>Reading attributes...</Text>
+        </CardBody>
+      </Card>
+    )
+  }
+
+  const checkIfAllMBeansHaveSameAttributes = (attributes: AttributeValues[]) => {
+    if (attributes.length <= 1) {
+      return true
+    }
+
+    const firstAttributeElements = attributes[0].length
+
+    if (!attributes.every(attribute => attribute.length === firstAttributeElements)) {
+      return false
+    }
+
+    return attributes.every(attribute =>
+      Object.keys(attribute).every(label => Object.keys(attributes[0]).includes(label)),
+    )
+  }
+
+  if (
+    attributes.some(attribute => Object.keys(attribute).length === 0) ||
+    attributes.some(attribute => Object.values(attribute).length === 0) ||
+    !checkIfAllMBeansHaveSameAttributes(attributes)
+  ) {
+    return <JmxContentMBeans />
```

Comment:
ok, so AttributeTable redispatches to JmxContentMBeans when the children are not of same kind. This flow looks a bit ugly to me but let's put up with it.

Reply:
As you need to fetch the data before making any decisions, this I found to be the better way right now to do this.

I'd like to do another refactor on to this that makes so the parent component is the one responsible for fetching the necessary data and then supplies it to the components below who only have to render, but I wanted to have something working first and then refactor later to a better design (nor that I didn't know if it'd be wanted so I decided to put that part in a separate PR later)

Reply:
yeah, I understand that we need to fetch attributes set proactively to make decisions at the earlier stage, so I also think we can leave it as-is for now. Let's fix the other parts and try to merge it first.